modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/IgniteCustomAssigmentsRules.java

```diff
@@ -97,6 +81,17 @@ private IgniteCustomAssigmentsRules(
             rules.add(type, rule);
         }
 
+        rule.addAll(YEAR_INTERVAL_TYPES);
+        rule.addAll(DAY_INTERVAL_TYPES);
+
+        // TINYINT is assignable from...
+        // SMALLINT is assignable from...
+        // INTEGER is assignable from...
+        // BIGINT is assignable from...
+        for (SqlTypeName type : EXACT_TYPES) {
+            rules.add(type, rule);
+        }
+
         // VARBINARY is assignable from...
         rule.clear();
         rule.addAll(BINARY_TYPES);
```

Comment:
conversion CHARACTER <-> BINARY is not allowed.

BTW, rules for BINARY and VARBINRY are exactly the same. Why not reuse the rule set? The same for [VAR-]CHAR